### ISSUE TITLE
typecheck: lacks support of import path with version suffix (gopkg.in/yaml.v3)

### ISSUE THREAD

**Issue by dolmen** [golangci/golangci-lint#4691]
### Welcome

- [X] Yes, I'm using a binary release within 2 latest major releases. Only such installations are supported.
- [X] Yes, I've searched similar issues on GitHub and didn't find any.
- [X] Yes, I've read the typecheck section of the FAQ (https://golangci-lint.run/welcome/faq/#why-do-you-have-typecheck-errors).
- [X] Yes, I've tried with the standalone linter if available (e.g., gocritic, go vet, etc.). (https://golangci-lint.run/usage/linters/)

### Description of the problem

The typecheck phase fails at resolving import path `gopkg.in/yaml.v3` as providing the symbol `yaml`. The Go 1.22 compiler has no problem with that import path.

### Version of golangci-lint

<details>

```console
$ golangci-lint --version
golangci-lint has version 1.57.2 built with go1.22.1 from 77a8601 on 2024-03-28T18:39:06Z
```

</details>

### A minimal reproducible example

```console
$ cat x.go
package main

import "gopkg.in/yaml.v3"

var _ = yaml.NewEncoder
$ golangci-lint run x.go
x.go:5:9: undefined: yaml (typecheck)
var _ = yaml.NewEncoder
        ^
```

**Comment by ldez** [comment]
Hello,

I recommend reading: https://golangci-lint.run/welcome/faq/#why-do-you-have-typecheck-errors

```console
$ go mod tidy                                   
go: finding module for package gopkg.in/yaml.v3
go: found gopkg.in/yaml.v3 in gopkg.in/yaml.v3 v3.0.1
$ golangci-lint run -v                           
INFO [config_reader] Config search paths: [./ /home/ldez/sources/go/src/github.com/golangci/sandbox /home/ldez/sources/go/src/github.com/golangci /home/ldez/sources/go/src/github.com /home/ldez/sources/go/src /home/ldez/sources/go /home/ldez/sources /home/ldez /home /] 
INFO [lintersdb] Active 6 linters: [errcheck gosimple govet ineffassign staticcheck unused] 
INFO [loader] Go packages loading at mode 575 (types_sizes|compiled_files|deps|exports_file|files|imports|name) took 64.139589ms 
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 76.312µs 
INFO [linters_context/goanalysis] analyzers took 0s with no stages 
INFO [runner] processing took 2.013µs with stages: max_same_issues: 488ns, nolint: 199ns, identifier_marker: 174ns, skip_dirs: 171ns, cgo: 138ns, filename_unadjuster: 126ns, invalid_issue: 122ns, max_from_linter: 70ns, fixer: 45ns, path_shortener: 42ns, skip_files: 41ns, exclude: 41ns, sort_results: 38ns, source_code: 38ns, severity-rules: 36ns, max_per_file_from_linter: 36ns, path_prettifier: 35ns, uniq_by_line: 35ns, autogenerated_exclude: 35ns, diff: 35ns, path_prefixer: 34ns, exclude-rules: 34ns 
INFO [runner] linters took 9.593813ms with stages: goanalysis_metalinter: 9.563524ms 
INFO File cache stats: 0 entries of total size 0B 
INFO Memory: 2 samples, avg is 31.3MB, max is 40.1MB 
INFO Execution took 75.863676ms     
```

**Comment by beautifulentropy** [comment]
FYI @dolmen I have the exact same issue in one of our private repos and right now. I prepared a minimal reproduction in an example repository. Much to my surprise, I cannot reproduce this issue outside of the repo where I experience it: https://github.com/beautifulentropy/golangci-lint-example/pull/3

**Comment by beautifulentropy** [comment]
> FYI @dolmen I have the exact same issue in one of our private repos and right now. I prepared a minimal reproduction in an example repository. Much to my surprise, I cannot reproduce this issue outside of the repo where I experience it: [beautifulentropy/golangci-lint-example#3](https://github.com/beautifulentropy/golangci-lint-example/pull/3)

Followup: the following change to our `.golangci.yml` did the trick:

```yaml
run:
  modules-download-mode: readonly
```

**Comment by ldez** [comment]
@beautifulentropy can you check the value of your `GOPROXY`?

**Comment by beautifulentropy** [comment]
> @beautifulentropy can you check the value of your `GOPROXY`?

Upon further inspection we actually found that someone had added `vendor/*` to the `.gitignore` for this repository. The vendor directory was still intact but of course it was missing `gopkg.in/yaml.v3`. Once we removed the entry for `vendor/*` and committed the result of another `go mod vendor` we were able to remove `modules-download-mode: readonly` from the `.golangci.yml` file and pass lints successfully.
